sklearn/linear_model/_linear_loss.py

```diff
@@ -8,6 +8,29 @@
 from ..utils.extmath import squared_norm
 
 
+def sandwich_dot(X, W):
+    """Compute the sandwich product X.T @ diag(W) @ X."""
+    # TODO: This "sandwich product" is the main computational bottleneck for solvers
+    # that use the full hessian matrix. Here, thread parallelism would pay-off the
+    # most.
+    # While a dedicated Cython routine could exploit the symmetry, it is very hard to
+    # beat BLAS GEMM, even thought the latter cannot exploit the symmetry, unless one
+    # pays the price of a taking square roots and implements
+    #    sqrtWX = sqrt(W)[: None] * X
```

Comment:
Note that exploiting symmetric is not the only reason why a dedicated sandwich product kernel would make sense.

This line above would trigger and read/write round trip between RAM and CPU of the size of `X` (when `X` is too large to fit in CPU cache which is typically the case of interest). When `n_samples >> n_features`, a dedicated fused sandwich product kernel would only have to:

- read `n_samples * (n_feature + 1) / 2` from RAM;
- write `n_features ** 2` to RAM.

while what you propose would:

- read `n_samples * (n_feature + 1)` from RAM,  # sqrtWX = sqrt(W)[: None] * X
- write `n_samples * n_feature` to RAM,  # sqrtWX = sqrt(W)[: None] * X
- read `n_samples * n_feature / 2` from RAM,  # np.dot(sqrtWX.T, sqrtWX)
- write `n_features ** 2` to RAM.  # np.dot(sqrtWX.T, sqrtWX)

Assuming that this kernel is memory bound, I would expect a ~3x speed-up from the fused kernel over the 2-step numpy code.

The problem is that writing an efficient blocked sandwich product kernel in Cython with OpenMP threading and hardware adapted SIMD vector instructions is far from trivial.

For CPU, https://github.com/Quantco/tabmat already presumably does that.

For GPU, something like https://github.com/openai/triton/ might be able to do it in a vendor agnostic way.

Reply:
I can remove some of that comment. I wanted to stress 2 facts:
- this is the cpu bottleneck
- Replacing it by some self written BLAS like function is a ludicrous undertaking (even tabmat is only faster when there are categoricals!). GEMM might be the algo where most human time was spent writing and optimizing it.

Reply:
Let’s not get off-topic too much.

Reply:
> Replacing it by some self written BLAS like function is a ludicrous undertaking (even tabmat is only faster when there are categoricals!). GEMM might be the algo where most human time was spent writing and optimizing it.

tabmat is near 4x faster than GEMM on dense numeric inputs according to:

- https://tabmat.readthedocs.io/en/latest/benchmarks.html#performance

This matches my memory bandwidth analysis above. My suggestion is not to change the implementation as part of this PR or in the near future but rather to improve the comment and converge on shared understanding of the remaining achievable performance improvements if we move away from GEMM to a dedicated sandwich product fused kernel (as the one implemented in tabmat).